[tensorexpr] Add BufLiveRange to obtain the liveness range of a buf in a

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66967
* #66452
* #67144
* #67143
* #66451
* __->__ #66450

block stmt

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

huiguoo committed:
[tensorexpr] Add a stmt recorder to obtain stmt PCs

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add a stmt recorder to obtain stmt Counters in AST"


A stmt counter is a vector of integers for stmt A: each integer represents the number of stmts
before A's ancestor or A in the block.

For example, "1:1:0" points to stmt "d[i] += e[i, j];" in the following code,

 c = a/b;
 for i
  d[i] = 0;
  for j
    d[i] += e[i, j];

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add a stmt recorder to obtain stmt Counters in AST"


A stmt counter is a vector of integers for stmt A: each integer represents the number of stmts
before A's ancestor or A in the block.

For example, "1:1:0" points to stmt "d[i] += e[i, j];" in the following code,

 c = a/b;
 for i
  d[i] = 0;
  for j
    d[i] += e[i, j];

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add a stmt recorder to obtain stmt Counters in AST"


A stmt counter is a vector of integers for stmt A: each integer represents the number of stmts
before A's ancestor or A in the block.

For example, "1:1:0" points to stmt "d[i] += e[i, j];" in the following code,

 c = a/b;
 for i
  d[i] = 0;
  for j
    d[i] += e[i, j];

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

huiguoo committed:
Update on "[tensorexpr] Add a stmt recorder to obtain stmt Counters in AST"


A stmt counter is a vector of integers for stmt A: each integer represents the number of stmts
before A's ancestor or A in the block.

For example, "1:1:0" points to stmt "d[i] += e[i, j];" in the following code,

 c = a/b;
 for i
  d[i] = 0;
  for j
    d[i] += e[i, j];

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add a stmt recorder to obtain stmt Counters in AST"


A stmt counter is a vector of integers for stmt A: each integer represents the number of stmts
before A's ancestor or A in the block.

For example, "1:1:0" points to stmt "d[i] += e[i, j];" in the following code,

 c = a/b;
 for i
  d[i] = 0;
  for j
    d[i] += e[i, j];

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

navahgar reviewed (commented):
Comments inline below.

Can you also add some tests for this?

huiguoo commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,220 @@ class ModifiesVarChecker : public IRVisitor {
+};
+
+// Visit Stmts and Record their PCs
+class StmtRecorder : public IRVisitor {
```
@navahgar I am thinking to replace `StmtCounter` with `StmtASTTrace`, and to replace `StmtRecorder` with `StmtASTTracer`. The new names sound more appropriate regarding how they are implemented (a stmt PC is implemented as a trace in the AST) but may not indicate as much about how they are used (that is they are used to denote the execution order of two stmts). Any thoughts?

huiguoo committed:
Update on "[tensorexpr] Add a stmt tracer to obtain stmt positions in AST; Add a"


class to obtain accesses to a buf

A stmt trace is a vector of integers for stmt A: each integer represents the number of stmts
before A's ancestor or A in the block.

For example, "1:1:0" points to stmt "d[i] += e[i, j];" in the following code,

 c = a/b;
 for i
  d[i] = 0;
  for j
    d[i] += e[i, j];

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,220 @@ class ModifiesVarChecker : public IRVisitor {
+  }
+
+  StmtCounter counter_;
+  bool flatten_ = false;
```
Why don't we know that S1 executes before S2? Would it be better if we simply specify which branch of If the stmt belong to? E.g. for
```
for {
  S1
  S2
  if {
    S3
    S4
  } else {
    S5
  }
}
```
S1 would have [0], S2 would have [1], S3 - [2, 0, 0], S4 - [2, 0, 1], S5 - [2, 1, 0].

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+ public:
+  BufAccesses(BufPtr b) : buf_(b) {}
+
+  std::vector<std::tuple<StmtPtr, AccMode, StmtTrace>> accesses() {
```
Nit: probably should be `std::vector<BufAccessInfo>`

huiguoo replied:
> Why don't we know that S1 executes before S2? Would it be better if we simply specify which branch of If the stmt belong to?

We know S1 executes before S2 but we don't know S3 executes before S4 because we'll flatten the if-stmt to one node in the AST. In this example, S1 - [0], S2 - [1], S3/S4/S5 - [2].
The reason we flatten if-stmt is that we can not differentiate if branches using a vector of integers: in current definition, an integer in the vector represents the stmt counter in a `Block`, and all stmts in a block will be executed from top to the bottom. That's why we can use the vector of integers to identify the execution order of buffer access stmts.
In the example, if S3 - [2, 0, 0], S5 - [2, 1, 0], then we will assume S5 executes after S3. One of the corresponding stmts is,

```
for {
  S1
  S2
  if {
    S3
    S4
  } 
  if { 
    S5
  }
}
```

We add extend `StmtTrace` to express cond stmts if there are use cases in the future.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+
+// Traverses the IR to identify all reads/writes to a buf, and their positions
+// in the AST which is represented as a StmtTrace
+using BufAccessInfo = std::tuple<StmtPtr, AccMode, StmtTrace>;
```
StmtTrace can be omitted from this tuple since we can always obtain it when we have StmtPtr. Generally it's better to pass as little (yet enough) information as possible to keep the API simpler.

ZolotukhinM replied:
Also, we already have two `AccessInfo` classes:
* https://github.com/pytorch/pytorch/blob/efdb17b984dc0624290d3a943d18df78ba5f55ed/torch/csrc/jit/tensorexpr/registerizer.h#L51
* https://github.com/pytorch/pytorch/blob/efdb17b984dc0624290d3a943d18df78ba5f55ed/torch/csrc/jit/tensorexpr/mem_dependency_checker.h#L38

I think we really need to first try using them, or merging them, or deleting some of them before adding a third one.

huiguoo replied:
>StmtTrace can be omitted from this tuple since we can always obtain it when we have StmtPtr.

I don't think so. Each `StmtPtr` is a node in the AST of the root stmt. We don't have their `StmtIndex` somewhere but have to visit the root stmt again to identify its position. Here, we have identified the position of all buf access stmts in one visit and are passing this info. I think it's better to re-compute their positions by visiting the root stmt multiple times, one time for each access.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+  }
+
+ private:
+  bool readBuffer(StmtPtr s) {
```
Nit: the name doesn't seem to be reflecting what the method does accurately. Should it be something like findBufReads?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+    return false;
+  }
+
+  bool writeBuffer(StmtPtr s) {
```
Ditto.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+    return false;
+  }
+
+  void insertAccesses(StmtPtr s) {
+    bool has_reads = readBuffer(s), has_writes = writeBuffer(s);
+    if (has_reads && has_writes) {
+      auto acc = std::make_tuple(s, AccMode::REDUCE, getStmtTrace());
+      accesses_.push_back(acc);
+      return;
+    }
```
Not all Stmts that have reads and writes to a buffer are reduces. If at some point this function would be used for a Block, it could start returning incorrect results. I realize it's not used for Blocks at this moment, but it's preferable to structure the code "defensively" to prevent such situations in future when possible. Would it be possible to refactor this code somehow to avoid that?

huiguoo replied:
Oh, I am trying to differentiate different buf accesses: buf read, buf write and buf read+write. `Reduce` operations both read and write the buf so I named the `read+write` mode with `REDUCE`. I can change it to `AccMode::BOTH`.

huiguoo commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+//  d[i] = 0;
+//  for j
+//    d[i] += e[i, j];
+class StmtTrace {
```
I think we need this class/wrapper if I understand you correctly. This class overwrites the comparison operators for `StmtTrace` and a bunch of utility functions such as `print`.

ZolotukhinM replied:
Yes, I see the methods we added to it. But I don't think any of them can't be replaced with a free standing function, we don't really need a class to print a vector.

huiguoo replied:
Yeah, it depends whether we want to encapsulate them to a class. I'm ok with either.

huiguoo commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+
+// Traverse the root stmt to identify the position of each stmt in the AST,
+// i.e., a StmtTrace
+class StmtTracer : public IRVisitor {
```
I understand your concerns. How about we use `StmtIndex` to replace `StmtTrace`,  and use `StmtIndexer` to replace `StmtTracer`?

ZolotukhinM replied:
Index/Indexer sound better to me!

ZolotukhinM commented:
After discussing this offline, I think the StmtIndexer should not be used for the memory-allocation optimization. It can be separately added as an individually useful tool, but in that case I suggest we add support for indexing into Ifs as well.

As for the memory planning, I believe we only need to look at statements at the outer level and don't need to look at nested stmts. This is because all our Allocate and Free statements are at that level. To illustrate it with example:
```
{
for i in 0..32:
  A[i] = i
for i in 0..32:
  B[i] = A[i]+1
  C[i] = i*i
for i in 0..32:
  D[i] = C[i] * 2
} // outputs: [B, D]
```

If we dont do any memory optimizations and place allocate/free naively, we would get:
```
{
Allocate(A)
Allocate(C)
for i in 0..32:
  A[i] = i
for i in 0..32:
  B[i] = A[i]+1
  C[i] = i*i
for i in 0..32:
  D[i] = C[i] * 2
Free(C)
Free(A)
} // outputs: [B, D]
```

However, If we try to insert PlacementAllocate here, it would need to go in between `B[i] = ...` and `C[i] = ...`, i.e. into the body of the second loop:
```
{
Allocate(A)
for i in 0..32:
  A[i] = i
for i in 0..32:
  B[i] = A[i]+1
  PlacementAllocate(A, C)
  C[i] = i*i
for i in 0..32:
  D[i] = C[i] * 2
Free(A)
} // outputs: [B, D]
```

That would lead to us having one Allocate for A, 32 PlacementAllocates for C, and one Free for A. While it would work correctly in the current implementation of PlacementAllocate lowering in our codegens, it definitely doesn't feel right - for a single Allocate we should have exactly one Free, and PlacementAllocate(A, B) should semantically be equivalent to "Free(A), Allocate(B)".

If we look only at the outer statement, we would correctly avoid using placement allocate in this case. I think that could not only lead to a more robust, but also to a much simpler implementation. What do you think?

huiguoo commented:
Thanks for pointing this out! The current implementation would produce wrong IR for the given example as it will reuse buffer A for C and insert PlacementAllocate at the global level. This is not correct and we should fix it.

A general solution is to identify the scope of intermediate buffers and only reuse buffers when we are sure that we can free them safely. For example,
```
1 for i in 0..64
2   for j in 0..32
3     c[j] = a[i, j] * b[i, j]
4   for j in 0..32
5     out1[i, j] = c[j]
6   for j in 0..32
7     d[j] = a[i, j] + b[i, j]
8   for j in 0..32
9     out2[i, j] = d[j]
```
a, b are inputs and out1, out2 are outputs. we can reuse buffer c for d after the following analysis: 1) where to free c: after Line 5; 2) where to allocate d: before Line 6; 3) does free c happen before allocate d: yes. So the stmt after memory planning should be,
```
1 for i in 0..64
2   Allocate(c)     
3   for j in 0..32
4     c[j] = a[i, j] * b[i, j]
5   for j in 0..32
6     out1[i, j] = c[j]
7   PlacementAllocate(c, d)
8   for j in 0..32
9     d[j] = a[i, j] + b[i, j]
10  for j in 0..32
11    out2[i, j] = d[j]
12. Free(c)  
``` 
We can also move the allocations to the global level to match pre-allocation, but once we have these info, we'll be able to release memory promptly instead of releasing them till all computations are done.

Go back to question 1: where to insert Free(c). We need to go over all accesses to C to determine what iterations index its accesses: in the example, only iteration `j` appear in its index expressions so we can safely insert Free(c) beyond j after the last access (Line 5, in the original stmt). 

We can also leave this to future work and land PRs in this stack first - it's already a large stack. For now, I can go with what you said, "look only at the outer statement for memory reuse". I'll also add index for if-stmts in StmtIndex.

ZolotukhinM commented:
Yeah, let's go with looking at the outer statements only. I believe currently we always generate Allocate/Free stmts in the global scope (for different reasons, but still), and in future if we decide to allow them to be inside loops this algorithm would work without any changes if we apply it to the inner block containing Allocate/Free stmts.

huiguoo committed:
Update on "[tensorexpr] Add a stmt tracer to obtain stmt positions in AST; Add a"


class to obtain accesses to a buf

A stmt trace is a vector of integers for stmt A: each integer represents the number of stmts
before A's ancestor or A in the block.

For example, "1:1:0" points to stmt "d[i] += e[i, j];" in the following code,

 c = a/b;
 for i
  d[i] = 0;
  for j
    d[i] += e[i, j];

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add a stmt indexer to obtain stmt positions in AST; Add a"


class to obtain accesses to a buf

A stmtIndex is a vector of integers for stmt A: each integer represents the number of stmts
before A's ancestor or A in the block.

For example, "1:1:0" points to stmt "d[i] += e[i, j];" in the following code,

 c = a/b;
 for i
  d[i] = 0;
  for j
    d[i] += e[i, j];

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

navahgar reviewed (approved):
Assuming you already addressed all of Mikhail's comments.

LGTM

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,194 @@ class ModifiesVarChecker : public IRVisitor {
+      stmt_index_.append(count);
+      s->accept(this);
+      stmt_index_.pop();
+      count++;
```
Why do we need `count`?

huiguoo replied:
We use `count` to count the number of stmts we have traversed in the block. Stmts in a block is organized as a list. We need a separate integer to track how many stmts have been traversed.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,194 @@ class ModifiesVarChecker : public IRVisitor {
+// i.e., a StmtIndex
+class StmtIndexer : public IRVisitor {
+ public:
+  StmtIndex getStmtIndex() {
```
Hm, so how exactly is this class used? How do I get indices for two sub-statements from a root statement? It seems that this class only has internal details of `BufAccesses` class and thus should be made part of that?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,194 @@ class ModifiesVarChecker : public IRVisitor {
+    return true;
+  }
+
+  bool operator<(StmtIndex& index) {
+    auto compare = index.getStmtIndex();
+    int size = stmt_index_.size() < compare.size() ? stmt_index_.size()
+                                                   : compare.size();
+    for (int i = 0; i < size; i++) {
+      if (stmt_index_.at(i) > compare.at(i)) {
+        return false;
+      }
+    }
+    return !(*this == index);
+  }
```
Since we're no longer looking into inner statements in mem planning, we're no longer using these comparisons, right? If so, I'd prefer getting rid of these methods and consider removing this class altogether.  getStmtIndexString method can be made a static free-standing function.

We need to optimize our code for (future) readers, not for (present) writers: i.e. even if the code we're adding now might look better with these classes, we need to take into account how difficult it would be to grasp it in future for people unfamiliar with it. That's why I generally prefer not to introduce wrappers over standard classes - we can assume that everyone knows API of std::vector, but we can't assume people know API of a custom class we've just added (as an example: can you tell what methods are there in `AccessInfo` without looking at the code?).

huiguoo committed:
Update on "[tensorexpr] Add BufAccesses to obtain access nodes to a buf"




Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

huiguoo committed:
Update on "[tensorexpr] Add BufAccesses to obtain access nodes to a buf"




Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

huiguoo committed:
Update on "[tensorexpr] Add BufAccesses to obtain access nodes to a buf"




Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

ZolotukhinM commented:
I'm still not quite sold on this to be honest. What problem does this class help to solve? Currently it looks like we're just gonna commit dead code. Originally it was used in the linear scan, but then we discovered that that approach was flawed and it is not used there anymore.

We can position it as an independently useful tool, but for that to be the case we need to demonstrate (e.g. in tests) at least how it can be used. Currently it only allows to compute indices, but there is nothing we can do with them afterwards. We should at least add a mirror function that would return a statement given an index, but I'd suggest moving this out from this stack as it is not required for the rest of the PRs.

huiguoo commented:
@ZolotukhinM The purpose of this PR is to collect access stmts to a buf, and for each access node we record its access mode (read, write or both?), and the index to this node from the root stmt. For example, 
```
For i
    A[i] = 0
    For j
        A[i] += B[i, j]
```
In the stmt above, there are two access nodes to buffer A, 1) access `A[i] = 0` with mode `WRITE` and index `0`; 2) access `A[i] += B[i, j]` with mode `BOTH` and index `1:0`. 
I understand that for the current implementation of memory planning, it doesn't fully use the metadata. But this class is self-contained: it makes sense to record such info when we collects buf accesses.
Even for the access mode (read/write/both), it's not used in the current implementation of memory planning. 
Are we removing all the metadata?

huiguoo committed:
Update on "[tensorexpr] Add BufLiveRange to obtain the liveness range of a buf in a"


block stmt

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

huiguoo committed:
Update on "[tensorexpr] Add BufLiveRange to obtain the liveness range of a buf in a"


block stmt

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,110 @@ class ModifiesVarChecker : public IRVisitor {
+ public:
+  BufLiveRange(BufPtr b) : buf_(b) {}
+
+  std::tuple<int32_t, int32_t> getLiveRange() {
+    return std::make_tuple(begin_, end_);
+  }
```
This probably needs to be private?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,110 @@ class ModifiesVarChecker : public IRVisitor {
+        to_count_ = false;
+        enable_count = true;
+      }
+      s->accept(this);
```
What's the point of using visitor here? Can we analyze `s` directly? (Essentially, can we call "updateRange" here instead of going through `::visit` calls? If we do this, we probably need to rename `updateRange`).

Also, I think this implementation might yield incorrect results if we have nested blocks (as visitor for inner blocks will rewrite `to_count_` and `cur_index_` values).

huiguoo replied:
`to_count_` cannot be rewritten by inner blocks because it's guarded by `enable_count`: `enable_count` is true only when we access the outermost loops. I have tests with nested blocks and they yield correct results.

Currently, `NodeFinder` is called (to update buf liveness range) when there are stmts like `store/external call`s. I agree that it would be cleaner to call `NodeFinder` for outer loops instead so we don't need to go through the stmts by ourselves. Will make the changes.

huiguoo replied:
For `updateRange`: it is short for `findBufAccessesAndUpdateBufLiveRange`. Do you have other name suggestions?

ZolotukhinM replied:
> For updateRange: it is short for findBufAccessesAndUpdateBufLiveRange. Do you have other name suggestions?

I personally would probably go with something generic in this case, like `processStmt`. Note that "range" is already implied there because it's a method of `BufLiveRange` class, so there is not much value in repeating it in the method name (and it even can be confusing as it could mean that there are several different ranges we work with).

ZolotukhinM replied:
> I agree that it would be cleaner to call NodeFinder for outer loops instead so we don't need to go through the stmts by ourselves. Will make the changes.

Ah, I see how it works. Yes, I think it can become much simpler and easier to follow after your changes!

ZolotukhinM commented on torch/csrc/jit/tensorexpr/analysis.h:
```diff
@@ -248,6 +248,110 @@ class ModifiesVarChecker : public IRVisitor {
+  }
+
+  void updateRange(StmtPtr s) {
+    bool has_reads = findBufReads(s), has_writes = findBufWrites(s);
```
Nit: I know that it was me who suggested this name in the beginning, but I think if we return `bool`, we should be using a name like "hasBufWrites". If the name starts with "find.." a common assumption is that it'll return a collection of items, not a boolean value.

huiguoo committed:
Update on "[tensorexpr] Add BufLiveRange to obtain the liveness range of a buf in a"


block stmt

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).

ZolotukhinM commented on test/cpp/tensorexpr/test_bufliverange.cpp:
```diff
@@ -0,0 +1,87 @@
+  //   for (int i = 0; i < 32; i++) {
+  //     a[i] = 0; // StmtIndex: 0:0
+  //     for (int j = 0; j < 32; j++) {
+  //       a[i] = (a[i]) + (b[i, j]); // StmtIndex: 0:1:0
```
Nit: "StmtIndex: 0:1:0" seems to remain from the old implementation

ZolotukhinM reviewed (approved):
LGTM!

huiguoo committed:
Update on "[tensorexpr] Add BufLiveRange to obtain the liveness range of a buf in a"


block stmt

Differential Revision: [D31557189](https://our.internmc.facebook.com/intern/diff/D31557189)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31557189).
